datafusion-examples/examples/memory_pool_tracking.rs

```diff
@@ -0,0 +1,191 @@
+
+    // Run a simple query to show it works
+    let results = ctx.sql("SELECT * FROM test").await?.collect().await?;
+    println!(
+        "✓ Query executed successfully. Found {} rows",
+        results.len()
+    );
+
+    println!("\n{}", "-".repeat(50));
+
+    // Error case: Create a runtime with low memory limit to trigger errors
+    println!("Error case: Triggering memory limit error with detailed error messages");
+
+    // Use a WITH query that generates data and then processes it to trigger memory usage
+    match ctx.sql("
+        WITH large_dataset AS (
+            SELECT
+                column1 as id,
+                column1 * 2 as doubled,
+                repeat('data_', 20) || column1 as text_field,
+                column1 * column1 as squared
+            FROM generate_series(1, 2000) as t(column1)
+        ),
+        aggregated AS (
+            SELECT
+                id,
+                doubled,
+                text_field,
+                squared,
+                sum(doubled) OVER (ORDER BY id ROWS BETWEEN 100 PRECEDING AND CURRENT ROW) as running_sum
+            FROM large_dataset
+        )
+        SELECT
+            a1.id,
+            a1.text_field,
+            a2.text_field as text_field2,
+            a1.running_sum + a2.running_sum as combined_sum
+        FROM aggregated a1
+        JOIN aggregated a2 ON a1.id = a2.id - 1
+        ORDER BY a1.id
+    ").await?.collect().await {
+        Ok(results) => panic!("Should not succeed! Yet got {} batches", results.len()),
+        Err(e) => {
+            println!("✓ Expected memory limit error during data processing:");
+            println!("Error: {e}");
+            /* Example error message:
+                Error: Not enough memory to continue external sort. Consider increasing the memory limit, or decreasing sort_spill_reservation_bytes
+                caused by
+                    Resources exhausted: Additional allocation failed with top memory consumers (across reservations) as:
+                    ExternalSorterMerge[3]#112(can spill: false) consumed 10.0 MB,
+                    ExternalSorterMerge[10]#147(can spill: false) consumed 10.0 MB,
+                    ExternalSorter[1]#93(can spill: true) consumed 69.0 KB,
+                    ExternalSorter[13]#155(can spill: true) consumed 67.6 KB,
+                    ExternalSorter[8]#140(can spill: true) consumed 67.2 KB.
+                Error: Failed to allocate additional 10.0 MB for ExternalSorterMerge[0] with 0.0 B already allocated for this reservation - 7.1 MB remain available for the total pool
+             */
+        }
+    }
+
+    println!("\nNote: The error message above shows which memory consumers");
+    println!("were using the most memory when the limit was exceeded.");
+
+    Ok(())
+}
+
+/// Example 2: Manual tracking with custom consumers
+///
+/// This shows how to manually create and use memory consumers with tracking.
```

Comment:
I am not sure this example adds much value as I don't expect people to generate their own consumers outside the context of writing an operator (which is covered by the other example)

Let's remove it from this PR for now